test/controllers/tasks_controller_test.rb

```diff
@@ -79,50 +76,145 @@
+      get edit_task_path(-1)
+
+      must_respond_with :redirect
+      must_redirect_to tasks_path
     end
   end
 
-  # Uncomment and complete these tests for Wave 3
   describe "update" do
-    # Note:  If there was a way to fail to save the changes to a task, that would be a great
-    #        thing to test.
     it "can update an existing task" do
-      skip
-      # Your code here
+      task_hash = {
+        task: {
+          name: "new task",
+          description: "new task description",
+          complete_by: Date.new(2019, 4, 20),
+        },
+      }
+
+      patch task_path(task), params: task_hash
+
+      new_task = Task.find_by(id: task.id)
+
+      expect(new_task.name).must_equal task_hash[:task][:name]
+      expect(new_task.description).must_equal task_hash[:task][:description]
+      expect(new_task.complete_by).must_equal task_hash[:task][:complete_by]
+
+      must_respond_with :redirect
+      must_redirect_to task_path(task)
     end
 
     it "will redirect to the root page if given an invalid id" do
-      skip
-      # Your code here
+      task_hash = {
+        task: {
+          name: "new task",
+          description: "new task description",
+          complete_by: Date.new(2019, 4, 20),
+        },
+      }
+      patch task_path(123456), params: task_hash
+
+      must_respond_with :redirect
+      must_redirect_to tasks_path
     end
   end
 
   # Complete these tests for Wave 4
   describe "destroy" do
-    # Your tests go here
+    it "removes the task from the database" do
+      task = Task.first
+
+      expect {
+        delete task_path(task)
+      }.must_change "Task.count", -1
 
+      must_respond_with :redirect
+      must_redirect_to tasks_path
+
+      destroyed_task = Task.find_by(id: task.id)
+      expect(destroyed_task).must_be_nil
+    end
+
+    it "returns a 404 if the task does not exist" do
+      task_id = 1337432
+
+      expect(Task.find_by(id: task_id)).must_be_nil
+
+      expect {
+        delete task_path(task_id)
+      }.wont_change "Task.count"
+
+      must_respond_with :not_found
+    end
   end
 
   # Complete for Wave 4
   describe "toggle_complete" do
-    # Your tests go here
+    it "Will change complete to true if false" do
+      new_task = Task.new(complete: false)
+      new_task.save
+
+      new_task = Task.find(new_task.id)
+      expect(new_task.complete).must_equal false
+
+      put task_path(new_task.id)
+
+      new_task = Task.find(new_task.id)
+      expect(new_task.complete).must_equal true
+    end
+
+    it "Will change complete to false if true" do
+      completed_task = Task.new(complete: true)
+      completed_task.save
+
+      expect(completed_task.complete).must_equal true
+
+      put task_path(completed_task.id)
+      completed_task.save
+
+      completed_task = Task.find(completed_task.id)
+      expect(completed_task.complete).must_equal false
+    end
+
+    it "returns a 404 if the task does not exist" do
+      task_id = 1337532
```

Comment:
Theoretically this could be a valid ID. It's better to choose something which could **never** be a valid ID.  Either destroy an existing Task and use it's ID, or use something like -1, or "invalid id" as the route parameter.